src/runtime_src/core/pcie/driver/linux/xocl/subdev/pmc.c

```diff
@@ -0,0 +1,237 @@
+/*
+ * Copyright (C) 2020 Xilinx, Inc. All rights reserved.
+ *
+ * Authors: David Zhang <davidzha@xilinx.com>
+ *
+ * This software is licensed under the terms of the GNU General Public
+ * License version 2, as published by the Free Software Foundation, and
+ * may be copied, distributed, and modified under those terms.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ */
+
+#include "../xocl_drv.h"
+
+#define	PMC_ERR(pmc, fmt, arg...)	\
+	xocl_err(&(pmc)->pmc_pdev->dev, fmt "\n", ##arg)
+#define	PMC_WARN(pmc, fmt, arg...)	\
+	xocl_warn(&(pmc)->pmc_pdev->dev, fmt "\n", ##arg)
+#define	PMC_INFO(pmc, fmt, arg...)	\
+	xocl_info(&(pmc)->pmc_pdev->dev, fmt "\n", ##arg)
+#define	PMC_DBG(pmc, fmt, arg...)	\
+	xocl_dbg(&(pmc)->pmc_pdev->dev, fmt "\n", ##arg)
+
+#define	PMC_ERR1_STATUS_MASK	(1 << 24)
+#define	PMC_ERR_OUT1_EN_MASK	(1 << 24)
+#define	PMC_POR1_EN_MASK	(1 << 24)
+#define	PMC_ERR_OUT1_MASK	0x20
+#define	PMC_ERR_OUT1_EN		0x24
+#define	PMC_POR1_MASK		0x40
+#define	PMC_POR1_EN		0x44
+
+#define	PL_TO_PMC_ERROR_SIGNAL_PATH_MASK	(1 << 0)
+
+enum {
+	PMC_IORES_INTR = 0,
+	PMC_IORES_MUX,
+	PMC_IORES_MAX,
+};
+
+struct xocl_iores_map pmc_res_map[] = {
+	{ RESNAME_PMC_INTR, PMC_IORES_INTR },
+	{ RESNAME_PMC_MUX, PMC_IORES_MUX },
+};
+
+struct pmc {
+	struct platform_device  *pmc_pdev;
+	void __iomem 		*pmc_base_address[PMC_IORES_MAX]; 
+	struct mutex 		pmc_lock;
+};
+
+static int pmc_enable_reset(struct platform_device *pdev)
+{
+	struct pmc *pmc = platform_get_drvdata(pdev);
+	void __iomem *pmc_intr, *pmc_mux;
+	u32 val;
+	int rc = 0;
+
+	mutex_lock(&pmc->pmc_lock);
+	/*
+	 * The pmc_intr register is a temporary workaround in driver, it will
+	 * be handled in CIPs and then removed from metadata.
+	 */
+	pmc_intr = pmc->pmc_base_address[PMC_IORES_INTR];
+	if (pmc_intr != NULL) {
+		val = XOCL_READ_REG32(pmc_intr);
+		if (val & PMC_ERR1_STATUS_MASK) {
+			val &= ~PMC_ERR1_STATUS_MASK;
+			XOCL_WRITE_REG32(val, pmc_intr); 
+		}
+
+		XOCL_WRITE_REG32(PMC_ERR_OUT1_EN_MASK, pmc_intr + PMC_ERR_OUT1_EN);
+		val = XOCL_READ_REG32(pmc_intr + PMC_ERR_OUT1_MASK);
+		if (val & PMC_ERR_OUT1_EN_MASK) {
+			PMC_ERR(pmc, "mask 0x%x for PMC_ERR_OUT1_MASK 0x%x "
+			    "should be 0.\n", PMC_ERR_OUT1_EN_MASK, val);
+			rc = -EINVAL;
```

Comment:
Nit: this seems to be a hardware misbehavior than a invalid argument? Would -EIO be more proper

Reply:
fixed